core/src/state.rs

```diff
@@ -80,6 +82,58 @@ pub struct LiveState {
     pub child_tree: bool,
 }
 
+impl LiveState {
+    /// Converts this `LiveState` into a `LiveState` for the previous block.
+    ///
+    /// Useful for opertations like when you want to execute a block, but also need the state of the
+    /// block *before* it.
+    pub async fn to_prev_block_live_state<Block: BlockT>(self) -> sc_cli::Result<LiveState>
+    where
+        <Block::Hash as FromStr>::Err: Debug,
+    {
+        // We want to execute the block `at`, therefore need the state of the block *before* it.
+        let at = self
+            .at
+            .clone()
+            .map(|s| hash_of::<Block>(s.as_str()))
+            .transpose()?;
+
+        // Get the block number requested by the user, or the current block number if they
+        // didn't specify one.
+        let rpc = ws_client(&self.uri).await?;
+        let number = ChainApi::<(), Block::Hash, Block::Header, ()>::header(&rpc, at)
+            .await
+            .map_err(rpc_err_handler)
+            .and_then(|maybe_header| maybe_header.ok_or("header_not_found").map(|h| *h.number()))?;
```

Comment:
Is the header not exposing its `parent` hash? Then you dont need to subtract one.

Reply:
Yeah that's a much better approach, thanks